content/2019/vim-and-the-shell/index.md

```diff
@@ -0,0 +1,227 @@
+
+```text
+$ cat fruits.txt | column -t | sort -nrk 2
+blueberries  15
+apples       7
+ananas       6
+oranges      5
+bananas      4
+```
+
+or, print how many fruits there are in total:
+
+```text
+$ cat fruits.txt | awk '{sum += $2} END { print sum }'
+37
+```
+
+You can filter and transform text in virtually any way using the command line, you can remove duplicate lines with the absolute poetry of a command `awk '!visited[$0]++'` [(for the curious)](https://opensource.com/article/19/10/remove-duplicate-lines-files-awk), change curly braces to parenthesizes with `tr '{}' '()'`, or delete the first three characters in each line with `cut -c3-`.
+
+The point is, command line text processing is *powerful*. It's hard to describe the beauty of executing a chain of commands and watching it, pipe after pipe, transform your text like a perfect symphony of streams; the magic of doing so much work with so little keystrokes. The fact, however, that you are a vim user makes me believe that you know this feeling very well.
+
+Talking about vim, wouldn't it be awesome if you could use advanced text processing tools, like `awk`, `tr`, `jq`, `bc`, as you edit your vim buffer?
+It probably doesn't come as a surprise to you when I tell you that that in fact, you can.
+
+## The Big Bang
+
+Vim integrates smoothly with external programs, the underlying shell, and the UNIX toolchain. Specifically, you can execute commands from vim, read the standard output of a command into your buffer, and use your buffer (or parts of it) as standard input to a shell program.
+
+In vim, the bang (`!`) symbol is used to interact with external commands.
+
+### Executing External Programs from VIM
+
+For example, you can run a shell program directly from vim's command line with:
+
+```vim
+:! {cmd}
+```
+This will show the standard output of __cmd__ in the bottom of the vim window ([`:h !:cmd`][bang]).
+
+But boy, we are your starting. You can use the bang in conjunction with the `:read` and `:write` commands to read from and write to external shell program executions.
+
+### Reading the Output of a Command Into Your Buffer
+
+For example, using 
+
+```vim
+:read !{cmd}
+```
+
+will execute __cmd__ and insert its standard output into your buffer below the cursor. You can also specify a range to indicate where the output of __cmd__ should be inserted:
+
+```vim
+:3read !curl --silent ifconfig.me
+```
+
+Will insert your public IP address at line 3 ([`:h read!`][read]).
+
+### Using Your Buffer as Input to a Command
+
+The other way works, too, you can execute a program with contents of your buffer as input:
+
+```vim
+:[range]write !{cmd}
+```
+
+will use the lines specified by __range__ (or the whole buffer if not specified) as standard input to __cmd__ and display its standard output below the window.
+[`:h w_c`][write]
+
+For example, if you want to sum up a column of numbers in your buffer you can just type:
+
+```vim
+:write !awk '{sum += $1} END { print sum }'
+```
+
```

Comment:
I would show how to do a *very* simple yet *very* common request: execute lines of code (Python, JavaScript, etc.):

```vim
" execute current line in the Node.js interpreter
:.write !node

" execute whole buffer in the Python interpreter
:%write !python
```

Reply:
good suggestion, added an example on running a visual selection through the python interpreter. Also didn't know that the interpreter read from std in :p